File: app/src/main/java/com/xmx/tango/module/calendar/CalendarActivity.java
package com.xmx.tango.module.calendar;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.os.Bundle;

import com.xmx.tango.R;
import com.xmx.tango.base.activity.BaseTempActivity;

import org.xutils.view.annotation.ContentView;
import org.xutils.view.annotation.ViewInject;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cn.aigestudio.datepicker.bizs.calendars.DPCManager;
import cn.aigestudio.datepicker.bizs.decors.DPDecor;
import cn.aigestudio.datepicker.cons.DPMode;
import cn.aigestudio.datepicker.views.DatePicker;

/**
 * Created by The_onE on 2017/6/25.
 */
@ContentView(R.layout.activity_calendar)
public class CalendarActivity extends BaseTempActivity {

    @ViewInject(R.id.datePicker)
    private DatePicker picker;

    Map<String, DateData> map = new HashMap<>();

    @Override
    protected void initView(Bundle savedInstanceState) {
        Date now = new Date();
        picker.setDate(now.getYear() + 1900, now.getMonth() + 1);
        picker.setTodayDisplay(true);
        picker.setMode(DPMode.NONE);

        List<String> checkIn = new ArrayList<>();
        List<String> info = new ArrayList<>();
        List<DateData> data = DateDataEntityManager.getInstance().selectAll();
        for (DateData item : data) {
            String key = item.year + "-" + item.month + "-" + item.date;
            checkIn.add(key);
            map.put(key, item);
            if (item.date != now.getDate()
                    || item.month != now.getMonth()+1
                    || item.year != now.getYear()+1900) {
                info.add(key);
            }
        }
        DPCManager.getInstance().setDecorBG(checkIn);
        DPCManager.getInstance().setDecorL(info);
        DPCManager.getInstance().setDecorT(info);
        DPCManager.getInstance().setDecorR(info);

        picker.setDPDecor(new DPDecor() {
            @Override
            public void drawDecorBG(Canvas canvas, Rect rect, Paint paint, String data) {
                super.drawDecorBG(canvas, rect, paint, data);
                paint.setColor(Color.GREEN);
                canvas.drawCircle(rect.centerX(), rect.centerY(), rect.width() / 2F, paint);
            }

            @Override
            public void drawDecorL(Canvas canvas, Rect rect, Paint paint, String data) {
                super.drawDecorL(canvas, rect, paint, data);
                DateData dateData = map.get(data);
                paint.setTextSize(20);
                canvas.drawText("学" + dateData.study, rect.centerX(), rect.centerY(), paint);
            }

            @Override
            public void drawDecorT(Canvas canvas, Rect rect, Paint paint, String data) {
                super.drawDecorT(canvas, rect, paint, data);
                DateData dateData = map.get(data);
                paint.setTextSize(20);
                canvas.drawText("任" + dateData.mission, rect.centerX(), rect.centerY(), paint);
            }

            @Override
            public void drawDecorR(Canvas canvas, Rect rect, Paint paint, String data) {
                super.drawDecorR(canvas, rect, paint, data);
                DateData dateData = map.get(data);
                paint.setTextSize(20);
                canvas.drawText("复" + dateData.review, rect.centerX(), rect.centerY(), paint);
            }
        });
    }

    @Override
    protected void setListener() {

    }

    @Override
    protected void processLogic(Bundle savedInstanceState) {

    }
}

File: app/src/main/java/com/xmx/tango/module/calendar/CalendarActivity.kt
package com.xmx.tango.module.calendar

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.os.Bundle

import com.xmx.tango.R
import com.xmx.tango.base.activity.BaseTempActivity

import cn.aigestudio.datepicker.bizs.calendars.DPCManager
import cn.aigestudio.datepicker.bizs.decors.DPDecor
import cn.aigestudio.datepicker.cons.DPMode
import kotlinx.android.synthetic.main.activity_calendar.*
import java.util.*

/**
 * Created by The_onE on 2017/6/25.
 * 打卡签到Activity
 */
class CalendarActivity : BaseTempActivity() {

    internal var map: MutableMap<String, DateData> = HashMap() // 日期与数据的映射

    override fun initView(savedInstanceState: Bundle?) {
        setContentView(R.layout.activity_calendar)
        // 初始化日历
        val c = Calendar.getInstance()
        datePicker.setDate(c.get(Calendar.YEAR), c.get(Calendar.MONTH) + 1)
        datePicker.setTodayDisplay(true)
        datePicker.setMode(DPMode.NONE)

        val checkIn = ArrayList<String>() // 打卡的日期列表（显示背景）
        val info = ArrayList<String>() // 除今天外打卡的日期列表（显示数据信息）
        // 从数据库中获取数据
        val data = DateDataEntityManager.selectAll()
        data?.apply {
            for (item in this) {
                // 生成日历读取的字符串
                val key = item.year.toString() + "-" + item.month + "-" + item.date
                checkIn.add(key)
                // 日期与数据的映射
                map.put(key, item)
                // 不是今天则显示数据信息
                if (item.date != c.get(Calendar.DATE)
                        || item.month != c.get(Calendar.MONTH) + 1
                        || item.year != c.get(Calendar.YEAR)) {
                    info.add(key)
                }
            }
        }
        // 设置日历显示信息
        DPCManager.getInstance().setDecorBG(checkIn)
        DPCManager.getInstance().setDecorL(info)
        DPCManager.getInstance().setDecorT(info)
        DPCManager.getInstance().setDecorR(info)
        // 设置日历信息的显示方式
        datePicker.setDPDecor(object : DPDecor() {
            override fun drawDecorBG(canvas: Canvas, rect: Rect, paint: Paint, data: String?) {
                super.drawDecorBG(canvas, rect, paint, data)
                // 为打卡的日期设置背景颜色
                if (map.containsKey(data)) {
                    paint.color = Color.GREEN
                    canvas.drawCircle(rect.centerX().toFloat(), rect.centerY().toFloat(), rect.width() / 2f, paint)
                }
            }

            override fun drawDecorL(canvas: Canvas, rect: Rect, paint: Paint, data: String?) {
                super.drawDecorL(canvas, rect, paint, data)
                // 为除今天外打卡的日期设置显示当天学习数
                if (map.containsKey(data)) {
                    map[data]?.apply {
                        paint.textSize = 20f
                        canvas.drawText("学" + this.study,
                                rect.centerX().toFloat(), rect.centerY().toFloat(), paint)
                    }
                }
            }

            override fun drawDecorT(canvas: Canvas, rect: Rect, paint: Paint, data: String?) {
                super.drawDecorT(canvas, rect, paint, data)
                // 为除今天外打卡的日期设置显示当天完成任务数
                if (map.containsKey(data)) {
                    map[data]?.apply {
                        paint.textSize = 20f
                        canvas.drawText("任" + this.mission,
                                rect.centerX().toFloat(), rect.centerY().toFloat(), paint)
                    }
                }
            }

            override fun drawDecorR(canvas: Canvas, rect: Rect, paint: Paint, data: String?) {
                super.drawDecorR(canvas, rect, paint, data)
                // 为除今天外打卡的日期设置显示当天复习数
                if (map.containsKey(data)) {
                    map[data]?.apply {
                        paint.textSize = 20f
                        canvas.drawText("复" + this.review, rect.centerX().toFloat(), rect.centerY().toFloat(), paint)
                    }
                }
            }
        })
    }

    override fun setListener() {

    }

    override fun processLogic(savedInstanceState: Bundle?) {

    }
}
